packages/lesswrong/components/hooks/useDynamicTableOfContents.ts

```diff
@@ -0,0 +1,61 @@
+import { useMemo } from "react";
+import { MIN_HEADINGS_FOR_TOC, ToCData, extractTableOfContents, getTocAnswers } from "../../lib/tableOfContents";
+import { PostWithCommentCounts, getResponseCounts, postGetCommentCountStr } from "../../lib/collections/posts/helpers";
+import { parseDocumentFromString } from "../../lib/domParser";
+
+type PostMinForToc = PostWithCommentCounts & {
+  question: boolean;
+  tableOfContentsRevision?: ToCData;
+  tableOfContents?: ToCData;
+};
+
+export const useDynamicTableOfContents = ({
+  html,
+  post,
+  answers,
+}: {
+  html: string | null;
+  post: PostMinForToc | null;
+  answers: CommentsList[];
+}): ToCData | null => {
+  return useMemo(() => {
+    const precalcuatedToc = post?.tableOfContentsRevision ?? post?.tableOfContents;
+    if (precalcuatedToc) {
+      return precalcuatedToc;
+    }
+
+    if (!html) {
+      return {
+        html,
+        sections: []
+      }
+    }
+
+    const { sections = [], html: tocHtml = null } =
+      extractTableOfContents(parseDocumentFromString(html)) ?? {};
+
+    // Always show the ToC for questions, to avoid layout shift when the answers load
+    if (sections.length > MIN_HEADINGS_FOR_TOC || post?.question) {
```

Comment:
I know DRY concerns are often a matter of taste among high level engineers, but I would still like to hear why it's not worth DRY-ing this.

Reply:
The basic reason I couldn't pull out this whole thing into a lib function is that in the server case we (conditionally) `await` fetching the comments and answers, and in the client case we just proceed with the data we have. A lib function would have to have to be async to allow this which would be a pain because we'd have to then make the hook and any component that uses it async. And we couldn't just eagerly fetch the comments and answers on the server because we do want to avoid fetching them if there are no headings in the doc.

That said, I have now pulled out a bit more of the fiddly logic into lib functions, see [this commit](https://github.com/ForumMagnum/ForumMagnum/pull/8971/files/7c49e09185f21d16e6ae7b01bba3007d910ab67c..d289bf58c4141f90baf1e304ca14c9a830ba4aa0)